dask/dataframe/io/parquet/arrow.py

```diff
@@ -337,6 +335,78 @@ def _need_fragments(filters, partition_keys):
     return bool(filtered_cols - partition_cols)
 
 
+def _filters_to_expression(filters, propagate_null=False, nan_is_null=True):
```

Comment:
I see why, but it's unfortunate that we need to patch a function this complex. Is there a way we could only use the patched version of `filters_to_expression` if we're applying a filter that needs patching?

Reply:
Yes, we could probably flatten the filter and look for "is", "is not", "!=", or "not in". However, I don't think that does much to simplify the patch (nor does it make the code easier to understand or maintain).

Reply:
Yeah, totally agree it won't make the patch any smaller. I was mostly concerned about drift between this patch and the corresponding function upstream in `pyarrow`. Though maybe we have sufficient test coverage to catch any drift in our upstream build before a `pyarrow` release happens

Reply:
I agree that drift is a concern. We had a similar plan for `write_to_dataset` three years ago, and divergence in supported file-naming options is now a *new* blocker (see: https://github.com/dask/dask/issues/9968).

Overall, I'm hopeful that pyarrow will agree with the proposed changes here (or adopt something similar that we can use). However, I'm pretty sure dask is the primary consumer of their `filters_to_expression` utility, and so I wouldn't consider it a tragedy if dask needed to take ownership of this logic longer-term.

Reply:
Okay, sounds good. I'll defer to your judgement here. Based on https://github.com/dask/dask/issues/9845#issuecomment-1416114411 for @jorisvandenbossche it sounds like this sort of change would be welcome on the `pyarrow` side